service/src/main/java/bio/terra/pipelines/service/QuotasService.java

```diff
@@ -53,4 +54,23 @@ public UserQuota getQuotaForUserAndPipeline(String userId, PipelinesEnum pipelin
     }
     return userQuota.get();
   }
+
+  /**
+   * This method updates the quota consumed for a given user and pipeline. It will return the
+   * updated user quota object.
+   *
+   * @param userQuota - the user quota object to update
+   * @param newQuotaConsumed - the quota consumed
+   * @return - the updated user quota object
+   */
+  public UserQuota updateQuotaConsumed(UserQuota userQuota, int newQuotaConsumed) {
+    if (newQuotaConsumed < 0) {
+      throw new InternalServerErrorException("Quota consumed must be greater than or equal to 0");
```

Comment:
it does get returned to the user

```
{
  "jobReport": {
    "id": "3fa85f64-5717-4562-0009-2c963f66afa6",
    "description": "fail if quota < 1000",
    "status": "FAILED",
    "statusCode": 500,
    "submitted": "2024-11-18T17:29:48.746984Z",
    "completed": "2024-11-18T17:32:01.323572Z",
    "resultURL": "http://localhost:8080/api/pipelineruns/v1/array_imputation/start/result/3fa85f64-5717-4562-0009-2c963f66afa6"
  },
  "errorReport": {
    "message": "Quota consumed must be greater than or equal to 0",
    "statusCode": 500,
    "causes": []
  },
  "pipelineRunReport": {
    "pipelineName": "array_imputation",
    "pipelineVersion": 0,
    "wdlMethodVersion": "ImputationBeagle_development_v0.0.1"
  }
}
```

maybe instead we should log errors and the string here should be somethign more generic like "Internal Error" ?

Reply:
hmm yeah i feel like we'll want a way to distinguish error messages that we want the user to see (like the next one) and hide the truly internal ones (like this one). i think we maybe have a ticket for that already: https://broadworkbench.atlassian.net/browse/TSPS-340 but i think that means that the one a few lines down (`InternalServerErrorException("Quota consumed cannot be greater than user quota");`) should maybe not be an InternalServerErrorException ?

Reply:
oh I see, this method shouldn't be called at all if the quota consumed is greater than user quota. so that's accurately an InternalServerErrorException.

I'm on board! no change needed now